app/api/queue.v2/application/Queue.ts

```diff
@@ -0,0 +1,110 @@
+import { Job } from '../contracts/Job';
+import { JobsDispatcher } from '../contracts/JobsDispatcher';
+import { QueueAdapter } from '../contracts/QueueAdapter';
+import { JobSerializer } from './JobSerializer';
+
+interface JobConstructor<T extends Job> {
+  new (...args: any[]): T;
+}
+
+interface Definition<T extends Job> {
+  constructorFn: JobConstructor<T>;
+  dependenciesFactory?: (namespace: string) => Promise<Partial<T>>;
+}
+
+interface QueueOptions {
+  lockWindow?: number;
+  namespace?: string;
+}
+
+const optionsDefaults: Required<QueueOptions> = {
+  lockWindow: 1,
+  namespace: '',
+};
+
+export class Queue implements JobsDispatcher {
+  private queueName: string;
+
+  private adapter: QueueAdapter;
+
+  private options: Required<QueueOptions>;
+
+  private definitions: Record<string, Definition<any>> = {};
+
+  private serializer: JobSerializer;
+
+  constructor(
+    queueName: string,
+    adapter: QueueAdapter,
+    serializer: JobSerializer,
+    options: QueueOptions = {}
+  ) {
+    this.queueName = queueName;
+    this.adapter = adapter;
+    this.serializer = serializer;
+    this.options = {
+      ...optionsDefaults,
+      ...options,
+    };
+  }
+
+  private async serializeJob(job: Job) {
+    return this.serializer.serialize(job, this.options.namespace);
+  }
+
+  private async deserializeJob(id: string, serialized: string) {
+    return this.serializer.deserialize(id, serialized, this.definitions);
+  }
+
+  private async ensureQueue() {
+    try {
+      await this.adapter.createQueueAsync({ qname: this.queueName, vt: this.options.lockWindow });
+    } catch (e) {
+      if (e.name !== 'queueExists') {
+        throw e;
+      }
+    }
+  }
+
+  async dispatch(job: Job): Promise<void> {
+    await this.ensureQueue();
+    await this.adapter.sendMessageAsync({
+      qname: this.queueName,
+      message: await this.serializeJob(job),
+    });
+  }
+
+  async peek() {
+    await this.ensureQueue();
+    const message = await this.adapter.receiveMessageAsync({ qname: this.queueName });
+    if ('id' in message) {
+      return this.deserializeJob(message.id, message.message);
+    }
+
+    return null;
+  }
+
+  async complete(job: Job) {
+    await this.ensureQueue();
+    await this.adapter.deleteMessageAsync({ qname: this.queueName, id: job.id! });
+  }
+
+  async progress(job: Job) {
+    await this.ensureQueue();
+    await this.adapter.changeMessageVisibilityAsync({
+      qname: this.queueName,
+      id: job.id!,
+      vt: job.lockWindow ?? this.options.lockWindow,
+    });
+  }
+
+  register<T extends Job>(
```

Comment:
This is indeed a bigger question that merits a bigger answer, but I'll still leave something here for the record :)
First things first, having the dependency factories in the class, in any form, is a violation of the dependency rule and a violation to the dependency injection principle. The Jobs are part of the application layer, they implement application logic, and depend on abstractions: for example, on a `RelationshipsDataSource`, not a `MongoRelationshipsDataSource`. Deciding what implementation to use is not part of its responsibilities, it happens in the outermost layer. This is not new, that is why we have factories. They are the composition root of the dependency injection pattern. This implementation is not different to that. The only difference is that it uses a different strategy for the injection: instead of constructor injection, it uses attribute injection, because we use the constructor for a different purpose in the jobs.
Why is the registration mechanism necessary? Is not to ensure that a certain queue only processes a certain type of job. That is not needed, and should be handled by not pushing jobs to the wrong queue. The job is instantiated by a uwazi server instance and put in a queue. You can't simply put a class instance into Redis, so we serialize it. When we deserialize it and try to reconstitute the instance (to have the data and the behavior), how would the worker (running on a separate process) know which class to instantiate? I don't mean the name, the name is actually serialized, but how to find the class code. That's what the registration is for: if you are looking for this job, this is the class and this is how to build it's dependencies.
Maybe you already noticed, but going back to "the worker runs on a separate process", that's why we need to tell it how to build dependencies through the registration: they are only going to be built when the job is going to be processed.